File: iOS/Objective-C/SampleObjC/CoarseRelocDemoViewController.m
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.
#import "CoarseRelocDemoViewController.h"
#import "Views/SensorStatusView.h"
#import "Models/LocationProviderSensorStatus.h"

/// Whether the "Access WiFi Information" capability is enabled.
/// If available, the MAC address of the connected Wi-Fi access point can be used
/// to help find nearby anchors.
/// Note: This entitlement requires a paid Apple Developer account.
const bool haveAccessWifiInformationEntitlement = false;

// Whitelist of Bluetooth-LE beacons used to find anchors and improve the locatability
// of existing anchors.
// Add the UUIDs for your own Bluetooth beacons here to use them with Azure Spatial Anchors.
#define KNOWN_BLUETOOTH_PROXIMITY_UUIDS @[ \
          @"61687109-905f-4436-91f8-e602f514c96d", \
          @"e1f54e02-1e23-44e0-9c3d-512eb56adec9", \
          @"01234567-8901-2345-6789-012345678903", \
      ]

// This demo creates and saves an anchor. It then locates all anchors near the device based on sensor data.
@implementation CoarseRelocDemoViewController
{
    IBOutlet SensorStatusView *sensorStatusView;
    ASAPlatformLocationProvider *locationProvider;
    ASACloudSpatialAnchorWatcher *nearDeviceWatcher;
    int numAnchorsFound;
}

-(void)onCloudAnchorCreated {
    _ignoreTaps = NO;
    [_feedbackControl setHidden:YES];
    dispatch_async(dispatch_get_main_queue(), ^(void){
        [self->sensorStatusView update];
    });
    [_button setTitle:@"Tap to start next Session & look for anchors near device" forState:UIControlStateNormal];
    _step = DemoStepLookForNearbyAnchors;
}

-(void)onNewAnchorLocated:(ASACloudSpatialAnchor*)cloudAnchor {
    _ignoreTaps = NO;
    [_feedbackControl setHidden:YES];
    NSString *message = [NSString stringWithFormat:@"%d anchor(s) found. Tap to stop watcher", ++self->numAnchorsFound];
    [_button setTitle:message forState:UIControlStateNormal];
    _step = DemoStepStopWatcher;
}

- (void)renderer:(id<SCNSceneRenderer>)renderer updateAtTime:(NSTimeInterval)time {
    [super renderer:renderer updateAtTime:time];
    dispatch_async(dispatch_get_main_queue(), ^(void){
        if (self->sensorStatusView != nil) {
            [self->sensorStatusView update];
        }
    });
}

- (void)buttonTap:(UIButton *)sender {
    if (_ignoreTaps){
        return;
    }
    switch (_step) {
        case DemoStepPrepare:
            [_button setTitle:@"Tap to start Session" forState:UIControlStateNormal];
            _step = DemoStepCreateCloudAnchor;
            [self createLocationProvider];
            break;
        case DemoStepCreateCloudAnchor:
            _ignoreTaps = YES;
            _currentlySavingAnchor = YES;
            _saveCount = 0;

            [self startSession];
            [self attachLocationProviderToSession];

            // When you tap on the screen, touchesBegan will call createLocalAnchor and create a local ARAnchor.
            // We will then put that anchor in the anchorVisuals dictionary with a key of "" and call CreateCloudAnchor when there is enough data for saving.
            // CreateCloudAnchor will call onCloudAnchorCreated when its async method returns to move to the next step.
            [_button setTitle:@"Tap on the screen to create an Anchor ☝️" forState:UIControlStateNormal];
            break;
        case DemoStepLookForNearbyAnchors:
            _ignoreTaps = YES;
            [self stopSession];
            [self startSession];
            [self attachLocationProviderToSession];

            // We will get a call to onLocateAnchorsCompleted which will move to the next step when the locate operation completes.
            [self lookForAnchorsNearDevice];
            break;
        case DemoStepStopWatcher:
            _step = DemoStepStopSession;
            [nearDeviceWatcher stop];
            nearDeviceWatcher = nil;
            [_button setTitle: @"Tap to stop Session and return to the main menu" forState:UIControlStateNormal];
            break;
        case DemoStepStopSession:
            [self stopSession];
            locationProvider = nil;
            [sensorStatusView setModel:nil];
            [self moveToMainMenu];
            return;
        default:
            assert(false);
            _step = 0;
            return;
    }
}

- (void)createLocationProvider {
    locationProvider = [[ASAPlatformLocationProvider alloc] init];

    // Register known Bluetooth beacons
    locationProvider.sensors.knownBeaconProximityUuids = KNOWN_BLUETOOTH_PROXIMITY_UUIDS;

    // Display the sensor status
    NSObject <SensorStatusModel> *sensorStatus = [[LocationProviderSensorStatus alloc] initForLocationProvider:locationProvider];
    [sensorStatusView setModel:sensorStatus];

    [self enableAllowedSensors];
}

- (void)enableAllowedSensors {
    if (locationProvider != nil) {
        ASASensorCapabilities *sensors = locationProvider.sensors;
        sensors.bluetoothEnabled = true;
        sensors.wifiEnabled = haveAccessWifiInformationEntitlement;
        sensors.geoLocationEnabled = true;
    }
}

- (void)attachLocationProviderToSession {
    _cloudSession.locationProvider = locationProvider;
}

- (void)lookForAnchorsNearDevice {
    ASANearDeviceCriteria *nearDevice = [[ASANearDeviceCriteria alloc] init];
    nearDevice.distanceInMeters = 8.0f;
    nearDevice.maxResultCount = 25;

    ASAAnchorLocateCriteria *criteria = [[ASAAnchorLocateCriteria alloc] init];
    criteria.nearDevice = nearDevice;
    nearDeviceWatcher = [_cloudSession createWatcher:criteria];

    [_button setTitle: @"Looking for anchors near device" forState:UIControlStateNormal];
}

@end

File: iOS/Swift/SampleSwift/ViewControllers/CoarseRelocDemoViewController.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

// This demo creates and saves an anchor. It then locates all anchors near the device based on sensor data.
class CoarseRelocDemoViewController: BaseViewController {

    @IBOutlet var sensorStatusView: SensorStatusView!

    /// Whether the "Access WiFi Information" capability is enabled.
    /// If available, the MAC address of the connected Wi-Fi access point can be used
    /// to help find nearby anchors.
    /// Note: This entitlement requires a paid Apple Developer account.
    private static let haveAccessWifiInformationEntitlement = false

    /// Whitelist of Bluetooth-LE beacons used to find anchors and improve the locatability
    /// of existing anchors.
    /// Add the UUIDs for your own Bluetooth beacons here to use them with Azure Spatial Anchors.
    public static let knownBluetoothProximityUuids = [
        "61687109-905f-4436-91f8-e602f514c96d",
        "e1f54e02-1e23-44e0-9c3d-512eb56adec9",
        "01234567-8901-2345-6789-012345678903",
    ]

    var locationProvider: ASAPlatformLocationProvider?

    var nearDeviceWatcher: ASACloudSpatialAnchorWatcher?
    var numAnchorsFound = 0

    override func onCloudAnchorCreated() {
        ignoreMainButtonTaps = false
        step = .lookForNearbyAnchors

        DispatchQueue.main.async {
            self.feedbackControl.isHidden = true
            self.mainButton.setTitle("Tap to start next Session & look for anchors near device", for: .normal)
        }
    }

    override func onNewAnchorLocated(_ cloudAnchor: ASACloudSpatialAnchor) {
        ignoreMainButtonTaps = false
        step = .stopWatcher

        DispatchQueue.main.async {
            self.numAnchorsFound += 1
            self.feedbackControl.isHidden = true
            self.mainButton.setTitle("\(self.numAnchorsFound) anchor(s) found! Tap to stop watcher.", for: .normal)
        }
    }

    override func renderer(_ renderer: SCNSceneRenderer, updateAtTime time: TimeInterval) {
        super.renderer(renderer, updateAtTime: time)

        DispatchQueue.main.async {
            self.sensorStatusView.update()
        }
    }

    @objc override func mainButtonTap(sender: UIButton) {
        if (ignoreMainButtonTaps) {
            return
        }

        switch (step) {
        case .prepare:
            mainButton.setTitle("Tap to start Session", for: .normal)
            step = .createCloudAnchor
            createLocationProvider()
        case .createCloudAnchor:
            ignoreMainButtonTaps = true
            currentlyPlacingAnchor = true
            saveCount = 0

            startSession()
            attachLocationProviderToSession()

            // When you tap on the screen, touchesBegan will call createLocalAnchor and create a local ARAnchor.
            // We will then put that anchor in the anchorVisuals dictionary with a special key and call CreateCloudAnchor when there is enough data for saving.
            // CreateCloudAnchor will call onCloudAnchorCreated when its async method returns to move to the next step.
            mainButton.setTitle("Tap on the screen to create an Anchor ☝️", for: .normal)
        case .lookForNearbyAnchors:
            ignoreMainButtonTaps = true
            stopSession()
            startSession()
            attachLocationProviderToSession()

            // We will get a call to onLocateAnchorsCompleted which will move to the next step when the locate operation completes.
            lookForAnchorsNearDevice()
        case .stopWatcher:
            step = .stopSession
            nearDeviceWatcher?.stop()
            nearDeviceWatcher = nil
            mainButton.setTitle("Tap to stop Session and return to the main menu", for: .normal)
        case .stopSession:
            stopSession()
            self.locationProvider = nil
            self.sensorStatusView.setModel(nil)
            moveToMainMenu()
        default:
            assertionFailure("Demo has somehow entered an invalid state")
        }
    }

    private func createLocationProvider() {
        locationProvider = ASAPlatformLocationProvider()

        // Register known Bluetooth beacons
        locationProvider!.sensors!.knownBeaconProximityUuids =
            CoarseRelocDemoViewController.knownBluetoothProximityUuids

        // Display the sensor status
        let sensorStatus = LocationProviderSensorStatus(for: locationProvider)
        sensorStatusView.setModel(sensorStatus)

        enableAllowedSensors()
    }

    private func enableAllowedSensors() {
        if let sensors = locationProvider?.sensors {
            sensors.bluetoothEnabled = true
            sensors.wifiEnabled = CoarseRelocDemoViewController.haveAccessWifiInformationEntitlement
            sensors.geoLocationEnabled = true
        }
    }

    private func attachLocationProviderToSession() {
        cloudSession!.locationProvider = locationProvider
    }

    private func lookForAnchorsNearDevice() {
        let nearDevice = ASANearDeviceCriteria()!
        nearDevice.distanceInMeters = 8.0
        nearDevice.maxResultCount = 25

        let criteria = ASAAnchorLocateCriteria()!
        criteria.nearDevice = nearDevice
        nearDeviceWatcher = cloudSession!.createWatcher(criteria)

        mainButton.setTitle("Looking for anchors near device...", for: .normal)
    }
}
